Replace remaining is<Attr> usage with dynamicDowncast<Attr>

annevk opened the pull request:
#### d2719a7661ab51053c2bdfab55cd0bbdc59cc7dd
<pre>
Replace remaining is&lt;Attr&gt; usage with dynamicDowncast&lt;Attr&gt;
<a href="https://bugs.webkit.org/show_bug.cgi?id=268375">https://bugs.webkit.org/show_bug.cgi?id=268375</a>

Reviewed by Chris Dumez.

* Source/WebCore/html/HTMLFrameOwnerElement.h:
(WebCore::HTMLFrameOwnerElement::protectedContentDocument const):
* Source/WebCore/inspector/InspectorNodeFinder.cpp:
(WebCore::InspectorNodeFinder::searchUsingDOMTreeTraversal):
(WebCore::InspectorNodeFinder::searchUsingXPath):
(WebCore::InspectorNodeFinder::searchUsingCSSSelectors):
* Source/WebCore/xml/XPathNodeSet.cpp:
(WebCore::XPath::sortBlock):
(WebCore::XPath::NodeSet::sort const):
(WebCore::XPath::findRootNode):
(WebCore::XPath::NodeSet::traversalSort const):

Canonical link: <a href="https://commits.webkit.org/273834@main">https://commits.webkit.org/273834@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9836f9cadb50c9520e3445c8141eb09d0a55279e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36826 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15761 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39093 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39466 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32986 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/38099 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18272 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12876 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31541 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37388 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13309 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32545 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11626 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/11638 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32833 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40716 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33409 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/33175 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/37551 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11919 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9734 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35671 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13578 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8344 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12312 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12820 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/xml/XPathNodeSet.cpp:
```diff
@@ -97,7 +97,7 @@ static void sortBlock(unsigned from, unsigned to, Vector<Vector<Node*>>& parentM
         for (unsigned i = sortedEnd; i < to; ++i) {
             Node* node = parentMatrix[i][0];
-            if (is<Attr>(*node) && downcast<Attr>(*node).ownerElement() == commonAncestor)
+            if (RefPtr attr = dynamicDowncast<Attr>(*node); attr && attr->ownerElement() == commonAncestor)
```
The RefPtr here is not really necessary.

cdumez commented on Source/WebCore/inspector/InspectorNodeFinder.cpp:
```diff
@@ -91,8 +91,8 @@ void InspectorNodeFinder::searchUsingDOMTreeTraversal(Node& parentNode)
                 m_results.add(node);
-            if (is<HTMLFrameOwnerElement>(downcast<Element>(*node)))
-                performSearch(downcast<HTMLFrameOwnerElement>(*node).contentDocument());
+            if (RefPtr frameOwner = dynamicDowncast<HTMLFrameOwnerElement>(*node))
```
The RefPtr here is not really necessary as we only call a trivial getter on frameOwner.

cdumez commented on Source/WebCore/inspector/InspectorNodeFinder.cpp:
```diff
@@ -91,8 +91,8 @@ void InspectorNodeFinder::searchUsingDOMTreeTraversal(Node& parentNode)
-            if (is<HTMLFrameOwnerElement>(downcast<Element>(*node)))
-                performSearch(downcast<HTMLFrameOwnerElement>(*node).contentDocument());
+            if (RefPtr frameOwner = dynamicDowncast<HTMLFrameOwnerElement>(*node))
+                performSearch(frameOwner->contentDocument());
```
However, this should call `performSearch(frameOwner->protectedContentDocument().get());`, adding a  `protectedContentDocument()` getter to HTMLFrameOwnerElement that returns a RefPtr.

cdumez commented on Source/WebCore/inspector/InspectorNodeFinder.cpp:
```diff
@@ -175,8 +175,8 @@ void InspectorNodeFinder::searchUsingXPath(Node& parentNode)
 
-        if (is<Attr>(*node))
-            node = downcast<Attr>(*node).ownerElement();
+        if (RefPtr attr = dynamicDowncast<Attr>(*node))
```
RefPtr is not really needed here.

cdumez commented on Source/WebCore/xml/XPathNodeSet.cpp:
```diff
@@ -159,8 +159,8 @@ void NodeSet::sort() const
         parentsVector.append(node);
-        if (is<Attr>(*node)) {
-            node = downcast<Attr>(*node).ownerElement();
+        if (RefPtr attr = dynamicDowncast<Attr>(*node)) {
```
RefPtr is not really needed here.

cdumez commented on Source/WebCore/xml/XPathNodeSet.cpp:
```diff
@@ -181,8 +181,8 @@ void NodeSet::sort() const
 {
-    if (is<Attr>(*node))
-        node = downcast<Attr>(*node).ownerElement();
+    if (RefPtr attr = dynamicDowncast<Attr>(*node))
```
RefPtr is not really needed here.

cdumez commented on Source/WebCore/xml/XPathNodeSet.cpp:
```diff
@@ -212,15 +212,15 @@ void NodeSet::traversalSort() const
-        for (const Attribute& attribute : element.attributesIterator()) {
-            RefPtr<Attr> attr = element.attrIfExists(attribute.name());
+        for (const Attribute& attribute : element->attributesIterator()) {
+            RefPtr<Attr> attr = element->attrIfExists(attribute.name());
```
You can drop `<Attr>` since you're touching this line.

cdumez reviewed (approved):
LGTM

annevk committed:
Replace remaining is<Attr> usage with dynamicDowncast<Attr>
https://bugs.webkit.org/show_bug.cgi?id=268375

Reviewed by Chris Dumez.

* Source/WebCore/html/HTMLFrameOwnerElement.h:
(WebCore::HTMLFrameOwnerElement::protectedContentDocument const):
* Source/WebCore/inspector/InspectorNodeFinder.cpp:
(WebCore::InspectorNodeFinder::searchUsingDOMTreeTraversal):
(WebCore::InspectorNodeFinder::searchUsingXPath):
(WebCore::InspectorNodeFinder::searchUsingCSSSelectors):
* Source/WebCore/xml/XPathNodeSet.cpp:
(WebCore::XPath::sortBlock):
(WebCore::XPath::NodeSet::sort const):
(WebCore::XPath::findRootNode):
(WebCore::XPath::NodeSet::traversalSort const):

Canonical link: https://commits.webkit.org/273834@main
